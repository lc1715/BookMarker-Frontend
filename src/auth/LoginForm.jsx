import { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import ShowAlert from '../common/ShowAlert';
import LoadSpinner from '../common/LoadSpinner';
//Material UI
import Container from '@mui/material/Container';
import Box from '@mui/material/Box';
import Paper from '@mui/material/Paper';
import TextField from '@mui/material/TextField';
import Button from '@mui/material/Button';
import Typography from '@mui/material/Typography';


/** Login form
 * 
 * Route is '/login'
 * 
 * Shows form and manages the update to formData state 
 * when values in input fields change.
 * 
 * On submission:
 * - calls login function prop
 * - redirects to homepage ('/')
 */

function LoginForm({ login }) {
    const [formData, setFormData] = useState({
        username: '',
        password: ''
    });
    const [formErrors, setFormErrors] = useState([]);
    const [loadingInfo, setLoadingInfo] = useState(false);

    const navigate = useNavigate();

    /** Handles login form submission:
     * -Calls the login function in App.jsx which will get user token
     * -If login successful and token received, returns { success: true }, and navigates to homepage
     * -Otherwise returns { success: false, err }
     */
    async function handleSubmit(evt, data) {
        evt.preventDefault();
        setLoadingInfo(true);

        let result = await login(data);

        if (result.success) {
            navigate('/');
        } else {
            setFormErrors(result.err);
        }
    }

    /**Update the formData state and render those changes in the input fields */
    function handleChange(evt) {
        const { name, value } = evt.target;
        setFormData(data => ({ ...data, [name]: value }));
    }

    if (loadingInfo) {
        return <LoadSpinner />
    }

    return (
        <form onSubmit={handleSubmit}>
            <Container maxWidth="sm" sx={{
                display: 'flex',
                alignItems: 'center',
                justifyContent: 'center',
                mt: { xs: 7, md: 10 },
                mb: 4
            }}>
                <Paper elevation={4}
                    sx={{
                        width: '90%',
                        backgroundColor: '#FCFBF4'
                    }}
                >
                    <Box sx={{ width: '100%', display: 'flex', justifyContent: 'center' }}>
                        <Typography sx={{
                            fontSize: { xs: '35px', md: '47px' },
                            textAlign: 'center',
                            mt: 5
                        }}>
                            Login Form
                        </Typography>
                    </Box>

                    <Box sx={{ mt: 4, width: '100%', display: 'flex', justifyContent: 'center' }}>
                        <TextField
                            name="username"
                            value={formData.username}
                            onChange={handleChange}
                            label="Username"
                            sx={{ width: '75%' }}
                        />
                    </Box>

                    <Box sx={{ mt: 4, width: '100%', display: 'flex', justifyContent: 'center' }}>
                        <TextField
                            type="password"
                            name="password"
                            value={formData.password}
                            onChange={handleChange}
                            label="Password"
                            sx={{ width: '75%' }}
                        />
                    </Box>

                    {formErrors.length ?
                        <Box sx={{ mt: 1, display: 'flex', alignItems: 'center', justifyContent: 'center' }}>
                            <ShowAlert type='error' messages={formErrors} />
                        </Box>
                        :
                        null
                    }

                    <Box sx={{ width: '100%', display: 'flex', justifyContent: 'center' }}>
                        <Button
                            variant="contained"
                            size="medium"
                            sx={{ mt: 4, width: '75%', height: 54, backgroundColor: '#cf8d86' }}
                            onClick={(evt) => (handleSubmit(evt, formData))}
                            type='submit'
                        >
                            Submit
                        </Button>
                    </Box>

                    <Box sx={{ width: '100%', display: 'flex', justifyContent: 'center' }}>
                        <Button
                            variant="text"
                            size="medium"
                            sx={{ py: 1, mt: 4.5, mb: 4.5, color: 'text.secondary' }}
                            onClick={(evt) => (handleSubmit(evt, { username: 'TestUser', password: 'Password123' }))}
                            type='submit'
                        >
                            Demo User
                        </Button>
                    </Box>
                </Paper>
            </Container >
        </form >
    )
};

export default LoginForm;
