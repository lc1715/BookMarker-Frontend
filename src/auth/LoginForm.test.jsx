import { it, expect } from 'vitest';
import { screen, render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom'
import LoginForm from './LoginForm';

it('renders LoginForm component', () => {
    render(
        <MemoryRouter>
            <LoginForm />
        </MemoryRouter>
    );
})

it('matches with snapshot', () => {
    const { asFragment } = render(
        <MemoryRouter>
            <LoginForm />
        </MemoryRouter>
    )

    expect(asFragment()).toMatchSnapshot();
})


it('should render the element containing the text Login Form', () => {
    render(
        <MemoryRouter>
            <LoginForm />
        </MemoryRouter>
    )

    expect(screen.getByText('Login Form')).toBeInTheDocument()
})
