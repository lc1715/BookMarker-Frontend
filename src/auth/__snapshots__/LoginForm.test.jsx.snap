// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches with snapshot 1`] = `
<DocumentFragment>
  <form>
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm css-qd6y3d-MuiContainer-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation4 css-18icxhl-MuiPaper-root"
        style="--Paper-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="MuiBox-root css-1ykdma4"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1r90siv-MuiTypography-root"
          >
            Login Form
          </p>
        </div>
        <div
          class="MuiBox-root css-9v94vn"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-w7xktn-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for=":r1:"
              id=":r1:-label"
            >
              Username
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r1:"
                name="username"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-81qg8w"
                >
                  <span>
                    Username
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-1b2y9pw"
        >
          <div
            class="MuiFormControl-root css-ve84sq-MuiFormControl-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="password"
            >
              Password
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-jupps9-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-1dune0f-MuiInputBase-input-MuiOutlinedInput-input"
                id="password"
                name="password"
                type="password"
                value=""
              />
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-elo8k2-MuiInputAdornment-root"
              >
                <button
                  aria-label="display password"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-oeq6yz-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                    data-testid="VisibilityIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5M12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5m0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"
                    />
                  </svg>
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-81qg8w"
                >
                  <span>
                    Password
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-1ykdma4"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-cpyu8g-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Submit
          </button>
        </div>
        <div
          class="MuiBox-root css-1ykdma4"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1btt4ro-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Demo User
          </button>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
